### ISSUE TITLE
TypeError: noop is not a constructor 

### ISSUE THREAD

**Issue by altunyurt** [lhorie/mithril.js#914]
Happens with 0.2.1+. I haven't inspected the main cause, yet noop seems not suitable as a fallback for the expected type.  

Error is thrown at 
https://github.com/lhorie/mithril.js/blob/next/mithril.js#L777

**Comment by isiahmeadows** [comment]
Try it in 0.2.2. The previous release was premature and buggy, and will see
a lot of active work to fix the bugs after Christmas.

**Comment by aliem** [comment]
I found the same issue with the following trivial code, release 0.2.3 from npm.

```
const App = window.App = {
    controller() {
        const onclick = () => window.open(getOAuthUrl());
        return {
            onclick,
        };
    },
    view(ctl) {
        return m('div', [
            m('button', 'Login', { onclick: ctl.onclick }),
        ]);
    },
};

document.addEventListener('DOMContentLoaded', () =>
                          m.mount(document.getElementById('app'), App));

```

The error is in https://github.com/lhorie/mithril.js/blob/next/mithril.js#L1369 `checkPrevented` method.

**Comment by tivac** [comment]
It's due to your usage of ES2015. `new App.controller()` fails in the same way. The shorthand `controller() {` notation is the equivalent of `controller : () => {`, and arrow functions aren't instantiable with `new`.

Use `controller : function() {` instead.

**Comment by barneycarroll** [comment]
@tivac that's definitely not the case — the prime advertised use case for the `{ method(){} }` shorthand is as class methods — they're meant to be `this`-bindable. [Check out this Babel REPL usage](http://babeljs.io/repl/#?evaluate=true&presets=es2015%2Cstage-3&experimental=true&loose=false&spec=false&code=const%20App%20%3D%20%7B%0A%20%20controller()%7B%0A%20%20%20%20this.initialized%20%3D%20true%0A%20%20%7D%0A%7D%0A%0Aconsole.log(%0A%20%20(%20new%20App.controller%20).initialized%0A)).

I've been using this boilerplate for components for ages, it's never let me down:

```javascript
export default {
  controller(){
    this.count = 0
  },

  view : ctrl =>
    m( 'p', ++count + ' draws' )
}
```

And `noop` itself is a vanilla function — there's no problem executing these. I can only assume you guys must have traced these errors to the wrong components, because the statement executes without error whichever reference resolves, [as seen in this snippet](http://babeljs.io/repl/#?evaluate=true&presets=es2015%2Cstage-3&experimental=true&loose=false&spec=false&code=var%20component%0Avar%20controller%0A%0Afunction%20noop()%7B%7D%0A%0Acomponent%20%20%3D%20%7B%7D%0Acontroller%20%3D%20new%20(component.controller%20%7C%7C%20noop)()%0A%0Aconsole.log(%20controller%20)%0A%0Acomponent%20%3D%20%7B%0A%20%20controller()%7B%7D%0A%7D%0A%0Acontroller%20%3D%20new%20(component.controller%20%7C%7C%20noop)()%0A%0Aconsole.log(%20controller%20)).

**Comment by barneycarroll** [comment]
@aliem can you try opening your Javascript debugger and inspect the value of `controller.component` when the error occurs? I think we must be missing part of the puzzle.

**Comment by tivac** [comment]
@barneycarroll https://jsbin.com/qiluda/3/edit?html,js,output

```
runner-3.35.11.min.js:1 error TypeError: App.controller is not a constructor
    at qiluda.js:17:1
2016-03-18 10:37:46.094 qiluda.js:18 Uncaught TypeError: App.controller is not a constructor
```

You sure about that?

**Comment by barneycarroll** [comment]
You won't believe this, but jsbin is actually blocked by the corporate network here.

Having said that, [this MDN reference says you're right: *Method definitions are not constructable*](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions#Method_definitions_are_not_constructable) and Babel is wrong. And actually, when I try doing it in the Chrome console, it throws! *Nice*.

So I've been giving people Babel-only solutions for ages. Sorry guys, scratch everything I've said!

*But they aren't arrow functions either.*

**Comment by tivac** [comment]
> But they aren't arrow functions either.

Fair enough, I was using that as a lazy shorthand because I couldn't find that MDN page.

**Comment by isiahmeadows** [comment]
I've run into this bug myself. And to be honest, it only *sometimes* calls it with `new`. The problem is, IIRC even this code fails:

```js
var compoent = {
  controller: function controller() {
    if (!(this instanceof controller)) {
      throw new TypeError("must be called with `new`")
    }
  },

  view: function () {
    return m("div")
  },
}

m.render(document.body, m(component))
```

I've tried to attack this twice, only to fail both times. :frowning: If you all can figure out a working solution, I'll merge it in an instant.

**Comment by tivac** [comment]
@isiahmeadows That code works fine, and after some digging through the mithril code I don't see how a controller would ever be executed w/o a `new` in front of it. Not to say it isn't possible, I'd just be very surprised. It's certainly not something we've ever run into while working in our large mithril codebase.

https://jsbin.com/haqake/1/edit?js,output

**Comment by isiahmeadows** [comment]
I know at least `component: class {}` didn't work last time I tried.

**Comment by barneycarroll** [comment]
@isiahmeadows I think there might be a misinterpretation of the nature of the problem here.

Constructors are especially nasty business because there is no such thing as a 'constructor' type check in Javascript. Traditionally, the static analysis pattern of to look to see whether the first letter of the function's `name` is a capital. That's the best we have. Otherwise it is just a function. To put it in other words: functions are black boxes (apart from `length`) and whether they are arrows, methods, classes, constructors, or any other (until recently informal) variety is information that can only be gleaned by invoking them.

So there is no procedural way to guard against this without extending the language. 

The reason I was so dismayed about @patcavit pointing out how wrong I was is that the *only* solution to this is to change application code to guard against it. It turns out that using 'method definitions' as controllers is impossible because of baked in language features (well, we could chuck `try{}catch(){}` into Mithril source, but like you said earlier, **performance**).

So sadly, the solution is what @tivac clearly explained earlier: don't do this. In other words: "User error: computer says NO".

**Comment by isiahmeadows** [comment]
@barneycarroll @tivac Which would you prefer: always or never? The docs would also need to be updated for the latter case, BTW.

**Comment by tivac** [comment]
@isiahmeadows Yes, a component cannot be a class. See #618 

A controller can be a `class` though, there's no issues there that I'm aware of. https://jsbin.com/viqawa/2/edit?js,output

As for this particular issue: changing mithril to support this would be a breaking change that I don't think makes any sense. I'm very against it.

Maybe the docs could make it clearer that `controllers` **must** be instantiable? I haven't read them in a while so it's possible it could be called out in a scarier way.

**Comment by isiahmeadows** [comment]
@tivac I was speaking about the controller, to clarify.

**Comment by lhorie** [comment]
As far as I know, there's no way to differentiate between an arrow/method and a class at call site, so I can't implement code to only call `new` for instantiatable constructs.

So I'd say the workaround is what @tivac suggested: use a regular `function () {}` instead

**Comment by isiahmeadows** [comment]
I have finally come up with a broken pen: http://codepen.io/isiahmeadows/pen/eZgwqP?editors=1011

Note: it requires ES6 to be enabled to work.

**Comment by isiahmeadows** [comment]
@lhorie Would you like me to file a bug on that pen? Or would you rather [this pen](http://codepen.io/isiahmeadows/pen/pyRMJK?editors=1011) to work?

**Comment by barneycarroll** [comment]
@isiahmeadows the outcome of this conversation is that we have discovered that new ES6 function variations are not valid as controller definitions. There is no library fix to decide or push — we just need to make the documentation clear about this.

**Comment by isiahmeadows** [comment]
@barneycarroll 

On Gitter, @lhorie [did seem to think this could be a valid way to rectify this](https://gitter.im/lhorie/mithril.js?at=56ec792fbb4a1731739b22df), while still allowing the inconsistency:

```js
function init(controller, args) {
  if ("prototype" in f) {
    return new (f.bind.apply(f, [undefined].concat(args)))()
  } else {
    var inst = {}
    var ret = f.apply(inst, args)
    return ret !== undefined ? ret : inst
  }
}
```

I agree that the documentation is misleading, but something has to be done. ES6 can't exactly be ignored without further propagating confusion among users. We can tell them until they're blue in the face, but some will still inevitably keep getting confused (most can't recall the entire documentation, and we still see HTML and basic JS mistakes reported as issues here). To be honest, the best fix would be to make this invalid:

```js
var Subcomponent = {
  view: function () {
    return m("div")
  },
}

var Component = {
  view: function () {
    return Subcomponent // Error
  },
}
```

It's very breaking and almost the nuclear option, but that change would make it much easier to fix this to something acceptable to *somebody* as well as implement a few other key proposals.

**Comment by barneycarroll** [comment]
@isiahmeadows good thinking, nice solution. [Here's a bin running through the different scenarios I can think of](http://jsbin.com/jiwoheh/edit?js,console,output).

The inevitable gotchas: 
* We're explicitly violating the generic expectations of method definitions (ie having a `this` binding to the object they were defined in) to do something which is more intuitive from a Mithril angle. This is the original problem, which your new initialiser 'solves'.
* Primitive return values are allowed when controllers aren't constructors. I can't see this being a problem – you'd have to be seriously confused to bind things to `this` and return a primitive.

How to document the instantiation contract might be too difficult to do in a way that isn't hugely overwrought and alienating to the average user – might be worth having a page in the wiki (to go along with the DOM gotchas I was talking about a while back).

Excellent pragmatism and clear thinking on your part though. I like the new `init` proposal a lot. My vote: :shipit:

**Comment by tivac** [comment]
I don't think this is nearly the issue you two do, but i still write ES5.

Want to see perf comparison to the existing, very unsurprising `new` behavior first

**Comment by barneycarroll** [comment]
@patcavit @isiahmeadows how about we branch this and try a different approach to the usual commit-to-next-and-release-with-a-load-of-other-stuff-while-waiting-for-the-bug-reports-to-come-in?

I suggest:
* Branch off (`aprototypical-controllers`?) with @isiahmeadows patch 
* Formalise a suite of test cases which expands upon my bin to provide every conceivable controller structure, add these to the branch
* Use the test outcomes w/ + w/out Babel to document current and proposed behaviours on the wiki for side by side comparison 
* Run comparative performance stress tests w eg threadit / monsterdb / whatever, record these in the same table as above
* Merge into next depending on cost-benefit-analysis of the above

**Comment by isiahmeadows** [comment]
@barneycarroll If you'd like, I can create another branch (hopefully with a better name than that monstrosity :wink:) for this. I'll try to periodically keep that in sync with `next`. Also, check out #994, so I'm not derailing this bug with that idea.

**Comment by aliem** [comment]
Thank you all for the discussion, and sorry for not being active in it. So, from now on I'll simply use a real functions and check your new branch (when it's up).

The shorthand function behavior is indeed unintuitive.

I personally don't feel controllers should be full fledged functions (or classes), from my perspective they could be just closures returning an object representing the state of the component.

**Comment by barneycarroll** [comment]
@aliem I completely agree with you. As discussed, the semantics of using a method definition as a constructor to create an instance of an object as `this`, which is referenced everywhere else as an explicit argument… is weird and difficult to reason about. I far prefer systems designed around explicit input and returns. But the brevity of

```javascript
{
  controller(){
    this.singleStateful = 'property'
  }
}
```

…is just too tempting, especially when writing demo code. It takes a lot of discipline to enforce the more rational but far more cruft laden:

```javascript
{
  controller: function(){
    return { singleStateful : 'property' }
  }
}
```

And Mithril is designed around that convenience.

I think the crucial task here — regardless of whether we make the patch or not — is to come up with documentation that explains your options and the rationale (such as it is) for the implementation as clearly as possible while elaborating on edge case expectations as necessary.

**Comment by isiahmeadows** [comment]
@aliem I would love to see your input on #994, BTW. :smile:

**Comment by nijikokun** [comment]
I know this is old but here is how I got around it, I don't use `this` so I don't really care about that, there are ways to add the `this` context throughout and you can figure it out yourself but here is your starting point:

```
app.class = (obj) => {
  let controller = obj.controller
  obj.controller = function () {
    return controller(...arguments)
  }

  return obj
}
```

How to use:

```
app.Views.Definitions = (obj) => {
  return app.class({
    controller () {},
    view (ctrl) {}
  })
}
```

**Comment by isiahmeadows** [comment]
@nijikokun 

If you run that in a browser with ES6 support, it's going to complain, because `controller` is always a class. I *could* trivially patch the component support to support calling functions that are missing a prototype, but it's low priority given the rewrite.

**Comment by nijikokun** [comment]
I am using it in the browser with no complaints

**Comment by isiahmeadows** [comment]
@nijikokun Oh...I misread it somehow... :worried:
